src/protocolsupport/protocol/packet/middleimpl/clientbound/play/v_pe/CraftingData.java

```diff
@@ -0,0 +1,143 @@
+		}
+		if (possibleStacks.length > 2) { //'large' ingredient set
+			NetworkItemStack firstType = ItemStackSerializer.remapItemToClient(connection.getVersion(), I18NData.DEFAULT_LOCALE, possibleStacks[0].cloneItemStack());
+			boolean allSameType = true;
+			//see if all ingredients are variants of the same type
+			for (NetworkItemStack ing : possibleStacks) {
+				NetworkItemStack thisType = ItemStackSerializer.remapItemToClient(connection.getVersion(), I18NData.DEFAULT_LOCALE, ing.cloneItemStack());
+				if(thisType.getTypeId() != firstType.getTypeId() || thisType.getAmount() != 1) {
+					allSameType = false;
+					break;
+				}
+			}
+			if (allSameType) {
+				NetworkItemStack wildStack = new NetworkItemStack();
+				wildStack.setTypeId(firstType.getTypeId());
+				wildStack.setLegacyData(-1); //wildcard stack
+				wildStack.setAmount(1);
+				possibleStacks = new NetworkItemStack[]{wildStack};
+			}
+		}
+		for (NetworkItemStack item : possibleStacks) {
+			LinkedList<NetworkItemStack> newItems = new LinkedList<>(items);
+			newItems.add(item);
+			if (tail.isEmpty()) {
+				consumer.accept(newItems); //final step
+			} else {
+				//recursively call with each new level of permutation
+				expandRecipe(tail, newItems, consumer);
+			}
+		}
+	}
+
+	@Override
+	public RecyclableCollection<ClientBoundPacketData> toData() {
+		recipesWritten = 0;
+		ByteBuf buffer = Unpooled.buffer();
+		for (Recipe recipe : recipes) {
+			if (recipe instanceof ShapelessRecipe) {
+				ShapelessRecipe shapeless = (ShapelessRecipe)recipe;
+
+				expandRecipe(shapeless.getIngredients(), ingredients -> {
+					addRecipeShapeless(buffer, shapeless.getResult(), ingredients);
+				});
+			} else if(recipe instanceof ShapedRecipe) {
+				ShapedRecipe shaped = (ShapedRecipe)recipe;
+
+				expandRecipe(shaped.getIngredients(), ingredients -> {
+					addRecipeShaped(buffer, shaped.getResult(), shaped.getWidth(), shaped.getHeight(), ingredients);
+				});
+			} else if(recipe instanceof SmeltingRecipe) {
+				SmeltingRecipe smelting = (SmeltingRecipe)recipe;
+				expandRecipe(new Ingredient[]{smelting.getIngredient()}, ingredients -> {
+					addRecipeFurnace(buffer, smelting.getResult(), ingredients.get(0));
+				});
+			}
+		}
+		ClientBoundPacketData craftPacket = ClientBoundPacketData.create(PEPacketIDs.CRAFTING_DATA);
+		VarNumberSerializer.writeVarInt(craftPacket, recipesWritten);
+		craftPacket.writeBytes(buffer);
+		buffer.release();
+		return RecyclableSingletonList.create(craftPacket);
+	}
+
+	protected void addRecipeShaped(ByteBuf to, NetworkItemStack output, int width, int height, List<NetworkItemStack> required) {
+		VarNumberSerializer.writeSVarInt(to, 1); //recipe type
+		VarNumberSerializer.writeSVarInt(to, width);
+		VarNumberSerializer.writeSVarInt(to, height);
+		for (NetworkItemStack stack : required) {
+			ItemStackSerializer.writeItemStack(to, connection.getVersion(), I18NData.DEFAULT_LOCALE, stack, stack == null || stack.getLegacyData() != -1);
```

Comment:
Why use the `I18NData.DEFAULT_LOCALE`? Get the player's locale using `cache.getAttributesCache().getLocale();`